### DISCUSSION TITLE
DeepAR Training failing due to invalid values in degrees of freedom (df) tensor. What is producing this behaviour?

### DISCUSSION THREAD

**Discussion by Serendipity31** [awslabs/gluonts#3044] [General]
## The Situation
I'm trying to tune DeepAREstimator hyperparameters, and every so often the call to `estimator.train()` fails in the middle of training due to a ValueError that is related to the _degrees of freedom tensor_ that is required by the output distribution. 

I have spent a long time trying to follow the traceback to figure out where this tensor gets created to try and figure out how it is that it could get populated with nan values, and I'm essentially none the wiser. 
- I can see multiple places in gluonTS and pytorch where distribution classes take degrees of freedom (i.e. `df`) as an input, but not where that input is created for the first time (or why it is a tensor and not an integer).
- I can also see `@classmethod` functions within distribution classes named `domain_map()` that modify this `df`, but I cannot see any place where `domain_map()` is called.

Can anyone please help??

### The ValueError
The full traceback is at the bottom of this post. Here is the ValueError on its own.
```{python}
"ValueError('Expected parameter df (Tensor of shape (72, 1)) of distribution Chi2() to satisfy the constraint 
GreaterThan(lower_bound=0.0), but found invalid values: \ntensor([     
[nan],\n        [nan],\n        [nan],\n        [nan],\n        [nan],\n        [nan],\n        [nan],\n        
[nan],\n        [nan],\n        [nan],\n        [nan],\n        [nan],\n        [nan],\n        [nan],\n        
[nan],\n        [nan],\n        [nan],\n        [nan],\n        [nan],\n        [nan],\n        [nan],\n        
[nan],\n        [nan],\n        [nan],\n        [nan],\n        [nan],\n        [nan],\n        [nan],\n        
[nan],\n        [nan],\n        [nan],\n        [nan],\n        [nan],\n        [nan],\n        [nan],\n        
[nan],\n        [nan],\n        [nan],\n        [nan],\n        [nan],\n        [nan],\n        [nan],\n        
[nan],\n        [nan],\n        [nan],\n        [nan],\n        [nan],\n        [nan],\n        [nan],\n        
[nan],\n        [nan],\n        [nan],\n        [nan],\n        [nan],\n        [nan],\n        [nan],\n        
[nan],\n        [nan],\n        [nan],\n        [nan],\n        [nan],\n        [nan],\n        [nan],\n        
[nan],\n        [nan],\n        [nan],\n        [nan],\n        [nan],\n        [nan],\n        [nan],\n        
[nan],\n        [nan]], grad_fn=<MulBackward0>)')
```
### Background
The following may be useful pieces of context:
- When training fails due to this error, it fails at some 'middle' epoch (e.g. 10 or 52 or 75 or 113), not at the very start

- The dimensions of this tensor match `batch_size`, and I've seen this error triggered with batch sizes both smaller and larger than the example shown above (which happens to be for `batch_size=72`)
- I've experienced this error both training a DeepAREstimator on its own and within the context of hyperparameter tuning with Optuna. 
- I don't think it has anything to do with Optuna, but when this happens within the context of hyperparameter tuning with Optuna, I have easy access to logs that show `train_loss` and `val_loss`. When this error happens, they are real numbers and they do not seem problematic (e.g. each might be somewhere between 1 and 2 and reasonably close to each other...not super small or super massive)
- My target series consists of real numbers, but they do have missing values. Sometimes there are quite a few of these in a row. If gluonTS does not implement a check to ensure that each selected window of length `batch_length` has at least one 'real' observation, it's possible that a batch within a particular target series and epoch _could_ exlusively consist of `nan` entries. It's possible this is relevant to this error.
- I use the default `DummyValueImputation(dummy_value=0.0)` as the imputation method. 
- Training does generate a RuntimeWarning (`RuntimeWarning: invalid value encountered in cast
  value = np.asarray(data[self.field], dtype=self.dtype)`). I discuss this here: [gluonts/issues/3025](https://github.com/awslabs/gluonts/issues/3025). I am able to reproduce this warning (scroll to the bottom of that thread), but have not gotten to the bottom of why it is happening. However, this RuntimeWarning happens every time I call `estimator.train()`, so these don't seem to be related. 

### Example  
Unfortunately, I don't have a reproducible example to provide. There are two reasons for this:
- I have not found a way to force it to generate the error on with a reproducible example, partially because I cannot see how `df` is created, which would allow me to create a `df` tensor that would directly trigger this error with a small reproducible example. 

- My training efforts with my real data have been interrupted several times with this error, but even with using `from lightning.pytorch import seed_everything` to set seeds for numpy, torch and python.random, I cannot actually reproduce this on command. I have, on one occaision managed to reproduce it a single time by re-running an Optuna study. However, afterwards I took that same dataset, and the same hyperparameter values, and trained the estimator outside of Optuna, and it did not produce the error (even using the same seed). I have also re-run an Optuna study after this error happened, and had the study run to completion without changing anything. It seems like there is some element of pseudo-randomness that is relevant that I'm not managing to control, and that this is making it harder to isolate the issue.

Even so, I've included an example of an estimator set up that did trigger this error recently. Although it's not reproducible, it shows the relevant settings.

```{python}
#Packages when training outside of Optuna
import torch
import pickle
import pandas as pd
from gluonts.dataset.split import split
from gluonts.evaluation import Evaluator
from lightning.pytorch import seed_everything
from gluonts.dataset.pandas import PandasDataset
from gluonts.torch.model.deepar import DeepAREstimator
from gluonts.torch.modules.loss import NegativeLogLikelihood
from gluonts.torch.distributions.studentT import StudentTOutput 

#Import dataset
with open("pythonPickles/dataset.pickle", "rb") as pickle_in:
  dataset = pickle.load(pickle_in)

#Function borrowed from the gluonTS tutorial for using Optuna
def dataentry_to_dataframe(entry):
  df = pd.DataFrame(
      data    = entry["target"], 
      columns = [entry.get("item_id")], 
      index   = pd.period_range(
          start   = entry["start"],
         periods = len(entry["target"]),
         freq    = entry["start"].freq
      ),
  )
  return df

#DataSet created with `from_long_dataframe()` from `gluonts.dataset.pandas` and then split into training
## and validation using split() and dataentry_to_dataframe() as defined above. This is also borrowed from
## the tutorial for integrating with Optuna
train, test_template = split(dataset, offset=-1)
validation = test_template.generate_instances(prediction_length=1)
validation_input = [entry[0] for entry in validation]
validation_label = [dataentry_to_dataframe(entry[1]) for entry in validation]

#Create list of cardinalities that are integers, not floats
cardinality_list = [round(x) for x in dataset.static_cardinalities.tolist()]

#Set up estimator
estimator = DeepAREstimator(
      freq                     = "M",
      prediction_length        = 1,
      context_length           = 1,
      nonnegative_pred_samples = True,
      num_parallel_samples     = 100,
      num_layers               = 3,   
      hidden_size              = 145,
      lr                       = 0.01,
      weight_decay             = 0.00000001,
      dropout_rate             = 0.17,
      batch_size               = 72,
      lags_seq                 = [1, 2, 3, 4, 5, 6, 7, 11, 12, 13],
      distr_output             = StudentTOutput(),
      loss                     = NegativeLogLikelihood(beta=0.01),
      num_feat_dynamic_real    = 9,
      num_feat_static_cat      = 33,
      cardinality              = cardinality_list,
      trainer_kwargs = {
          "max_epochs" : 200,
          "deterministic": True
      },
    )

seed_everything(42, workers=True)
predictor   = estimator.train(training_data=train, validation_data=test_template.dataset, cache_data=True)
```
### Full Traceback
Here is the full traceback
```{pyton}
Traceback (most recent call last):
  File "y:\shared\directory\Lib\site-packages\optuna\study\_optimize.py", line 200, in _run_trial
    value_or_values = func(trial)
                      ^^^^^^^^^^^
  File "Y:\shared\directory\deeparanalysis\pythonScripts\HP Tuning Scripts\model.py", line 329, in __call__
    predictor   = estimator.train(
                  ^^^^^^^^^^^^^^^^
  File "y:\shared\directory\Lib\site-packages\gluonts\torch\model\estimator.py", line 248, in train
    return self.train_model(
           ^^^^^^^^^^^^^^^^^
  File "y:\shared\directory\Lib\site-packages\gluonts\torch\model\estimator.py", line 211, in train_model
    trainer.fit(
  File "y:\shared\directory\Lib\site-packages\lightning\pytorch\trainer\trainer.py", line 545, in fit
    call._call_and_handle_interrupt(
  File "y:\shared\directory\Lib\site-packages\lightning\pytorch\trainer\call.py", line 44, in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "y:\shared\directory\Lib\site-packages\lightning\pytorch\trainer\trainer.py", line 581, in _fit_impl
    self._run(model, ckpt_path=ckpt_path)
  File "y:\shared\directory\Lib\site-packages\lightning\pytorch\trainer\trainer.py", line 990, in _run
    results = self._run_stage()
              ^^^^^^^^^^^^^^^^^
  File "y:\shared\directory\Lib\site-packages\lightning\pytorch\trainer\trainer.py", line 1036, in _run_stage
    self.fit_loop.run()
  File "y:\shared\directory\Lib\site-packages\lightning\pytorch\loops\fit_loop.py", line 202, in run
    self.advance()
  File "y:\shared\directory\Lib\site-packages\lightning\pytorch\loops\fit_loop.py", line 359, in advance
    self.epoch_loop.run(self._data_fetcher)
  File "y:\shared\directory\Lib\site-packages\lightning\pytorch\loops\training_epoch_loop.py", line 136, in run
    self.advance(data_fetcher)
  File "y:\shared\directory\Lib\site-packages\lightning\pytorch\loops\training_epoch_loop.py", line 240, in advance
    batch_output = self.automatic_optimization.run(trainer.optimizers[0], batch_idx, kwargs)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "y:\shared\directory\Lib\site-packages\lightning\pytorch\loops\optimization\automatic.py", line 187, in run
    self._optimizer_step(batch_idx, closure)
  File "y:\shared\directory\Lib\site-packages\lightning\pytorch\loops\optimization\automatic.py", line 265, in _optimizer_step
    call._call_lightning_module_hook(
  File "y:\shared\directory\Lib\site-packages\lightning\pytorch\trainer\call.py", line 157, in _call_lightning_module_hook
    output = fn(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^
  File "y:\shared\directory\Lib\site-packages\lightning\pytorch\core\module.py", line 1282, in optimizer_step
    optimizer.step(closure=optimizer_closure)
  File "y:\shared\directory\Lib\site-packages\lightning\pytorch\core\optimizer.py", line 151, in step
    step_output = self._strategy.optimizer_step(self._optimizer, closure, **kwargs)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "y:\shared\directory\Lib\site-packages\lightning\pytorch\strategies\strategy.py", line 230, in optimizer_step
    return self.precision_plugin.optimizer_step(optimizer, model=model, closure=closure, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "y:\shared\directory\Lib\site-packages\lightning\pytorch\plugins\precision\precision_plugin.py", line 117, in optimizer_step
    return optimizer.step(closure=closure, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "y:\shared\directory\Lib\site-packages\torch\optim\optimizer.py", line 373, in wrapper
    out = func(*args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^
  File "y:\shared\directory\Lib\site-packages\torch\optim\optimizer.py", line 76, in _use_grad
    ret = func(self, *args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "y:\shared\directory\Lib\site-packages\torch\optim\adam.py", line 143, in step
    loss = closure()
           ^^^^^^^^^
  File "y:\shared\directory\Lib\site-packages\lightning\pytorch\plugins\precision\precision_plugin.py", line 104, in _wrap_closure
    closure_result = closure()
                     ^^^^^^^^^
  File "y:\shared\directory\Lib\site-packages\lightning\pytorch\loops\optimization\automatic.py", line 140, in __call__
    self._result = self.closure(*args, **kwargs)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "y:\shared\directory\Lib\site-packages\torch\utils\_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "y:\shared\directory\Lib\site-packages\lightning\pytorch\loops\optimization\automatic.py", line 126, in closure
    step_output = self._step_fn()
                  ^^^^^^^^^^^^^^^
  File "y:\shared\directory\Lib\site-packages\lightning\pytorch\loops\optimization\automatic.py", line 315, in _training_step
    training_step_output = call._call_strategy_hook(trainer, "training_step", *kwargs.values())
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "y:\shared\directory\Lib\site-packages\lightning\pytorch\trainer\call.py", line 309, in _call_strategy_hook
    output = fn(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^
  File "y:\shared\directory\Lib\site-packages\lightning\pytorch\strategies\strategy.py", line 382, in training_step
    return self.lightning_module.training_step(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "y:\shared\directory\Lib\site-packages\gluonts\torch\model\deepar\lightning_module.py", line 74, in training_step
    train_loss = self.model.loss(
                 ^^^^^^^^^^^^^^^^
  File "y:\shared\directory\Lib\site-packages\gluonts\torch\model\deepar\module.py", line 565, in loss
    distr = self.output_distribution(params, scale)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "y:\shared\directory\Lib\site-packages\gluonts\torch\model\deepar\module.py", line 357, in output_distribution
    return self.distr_output.distribution(sliced_params, scale=scale)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "y:\shared\directory\Lib\site-packages\gluonts\torch\distributions\distribution_output.py", line 137, in distribution
    distr = self._base_distribution(distr_args)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "y:\shared\directory\Lib\site-packages\gluonts\torch\distributions\distribution_output.py", line 114, in _base_distribution
    return self.distr_cls(*distr_args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "y:\shared\directory\Lib\site-packages\gluonts\torch\distributions\studentT.py", line 39, in __init__
    super().__init__(
  File "y:\shared\directory\Lib\site-packages\torch\distributions\studentT.py", line 61, in __init__
    self._chi2 = Chi2(self.df)
                 ^^^^^^^^^^^^^
  File "y:\shared\directory\Lib\site-packages\torch\distributions\chi2.py", line 25, in __init__
    super().__init__(0.5 * df, 0.5, validate_args=validate_args)
  File "y:\shared\directory\Lib\site-packages\torch\distributions\gamma.py", line 58, in __init__
    super().__init__(batch_shape, validate_args=validate_args)

File "y:\shared\directory\Lib\site-packages\torch\distributions\distribution.py", line 68, in __init__
    raise ValueError(
ValueError: Expected parameter df (Tensor of shape (72, 1)) of distribution Chi2() to satisfy the constraint GreaterThan(lower_bound=0.0), but found invalid values:
tensor([[nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan],
        [nan]], grad_fn=<MulBackward0>)
```

**Comment by Serendipity31** [comment]
## Update  
I'm still trying to figure out what is going on here. I've found an iteration that seems to trigger this issue and have managed to go through it a number of times. I've narrowed down the problem.

The input data are not problematic. There are no `nan` values in the input data. The imputation seems to work as expected. the `ObservedValueIndicator` is created as expected. All seems normal up until `self.embedder` is called as a part of `prepare_rnn_input()`. This is where `nan` values suddenly appear in the data. 

Specifically, in [line 251](https://github.com/awslabs/gluonts/blob/536465d5fa979d62da0c88837d949067137d9697/src/gluonts/torch/model/deepar/module.py#L251) of deepar/module.py. This line is within `prepare_rnn_input()`, the definition of which is reproduced below in full.

```
def prepare_rnn_input(
        self,
        feat_static_cat: torch.Tensor,
        feat_static_real: torch.Tensor,
        past_time_feat: torch.Tensor,
        past_target: torch.Tensor,
        past_observed_values: torch.Tensor,
        future_time_feat: torch.Tensor,
        future_target: Optional[torch.Tensor] = None,
    ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor,]:
        context = past_target[..., -self.context_length :]
        observed_context = past_observed_values[..., -self.context_length :]

        input, _, scale = self.scaler(context, observed_context)
        future_length = future_time_feat.shape[-2]
        if future_length > 1:
            assert future_target is not None
            input = torch.cat(
                (input, future_target[..., : future_length - 1] / scale),
                dim=-1,
            )
        prior_input = past_target[..., : -self.context_length] / scale

        lags = lagged_sequence_values(
            self.lags_seq, prior_input, input, dim=-1
        )

        time_feat = torch.cat(
            (
                take_last(past_time_feat, dim=-2, num=self.context_length - 1),
                future_time_feat,
            ),
            dim=-2,
        )

        embedded_cat = self.embedder(feat_static_cat) #*****nan values appear here!*****
        static_feat = torch.cat(
            (embedded_cat, feat_static_real, scale.log()),
            dim=-1,
        )
        expanded_static_feat = unsqueeze_expand(
            static_feat, dim=-2, size=time_feat.shape[-2]
        )

        features = torch.cat((expanded_static_feat, time_feat), dim=-1)

        return torch.cat((lags, features), dim=-1), scale, static_feat
```

The first part of this return (i.e. `torch.cat((lags, features), dim=-1)`) containing `nan`, in turn, means that `rnn_input` (shown below from[ line 317](https://github.com/awslabs/gluonts/blob/536465d5fa979d62da0c88837d949067137d9697/src/gluonts/torch/model/deepar/module.py#L317) of deepar/module.py), contains `nan` values. 
```
rnn_input, scale, static_feat = self.prepare_rnn_input(
            feat_static_cat,
            feat_static_real,
            past_time_feat,
            past_target,
            past_observed_values,
            future_time_feat,
            future_target,
        )
```
Following on from line 317, in [line 327](https://github.com/awslabs/gluonts/blob/536465d5fa979d62da0c88837d949067137d9697/src/gluonts/torch/model/deepar/module.py#L327C9-L330C61), this `rnn_input` is then passed to in `self.rnn()`. In turn, this ensures that the first object returned from `self.rnn()` (i.e. `output`) is full of `nan`. The object `output` is subsequently an input to the calculation of `params`, shown in the code below:
```
output, new_state = self.rnn(rnn_input)

params = self.param_proj(output)
return params, scale, output, static_feat, new_state
```
The `params` object contains `df`, but it's `nan` (because `output` is full of `nan`). Incidentally, the mean and standard deviation are also `nan`. Consequently, when the check is performed to see if the argument constraint for `df` is met, that check fails. 

In my case, the distribution is a StudentT distribution, and the constraint seems to get checked in the `__init__` of the pytorch `Distribution` class (see [line 68](https://github.com/pytorch/pytorch/blob/ff7c06a01bec97c897374a685f33bb7253df11ec/torch/distributions/distribution.py#L68)). 

So the root of the problem isn't _really_ to do with `df`... the root of the problem is something to do with `self.embedder()`. It's only that the first check that happens after `self.embedder()` that could catch the appearance of `nan` is the argument constraint related to `df` in the `Chi2` distribution created as a part of the `StudentT` output distribution. 

## Questions
- Why would `self.embedder()` return `nan`? 

In this case, `batch_size` = 16, so there are 16 `feat_static_cat` tensors. The first is: 
[4, 6, 326, 6, 22, 61, 17, 36, 53, 2, 76, 1, 0, 8, 27, 45, 70, 1, 1, 1, 1, 0, 0, 4, 3, 4, 0, 2, 6, 45, 2, 13, 5]. 

Nothing in here appears to be abnormal or odd or problematic. All 16 such tensors in this case are of a type like this. All seem normal. There is nothing obviously problematic about the values in any of them. 

- Why do the values of in `embedded_cat` affect the parameters of the distribution? 

I guess this is probably more of a conceptual question, but the answer isn't obvious (to me).

**Comment by santoshkumarradha** [comment]
Running into same issue.

**Reply by Serendipity31** [reply]
@santoshkumarradha - Apologies. I didn't see that you'd replied here back in June. I am not 100% sure that I got to the bottom of this, but I can share with you what I did in case it helps you in anyway.  

After thinking on this problem for a while, I developed a hunch that this related to the exploding/vanishing gradients problem (and very probably the vanishing gradient part of this problem). Some of my static categorical variables had a low cardinality (i.e. not many values a series could take with respect to that variable). Others had quite a few, but the distribution across the full dataset was very skewed (i.e. many series being assigned to one category and just a few to the other categories within a variable). This made me wonder if the embedding dimensions were too big, in practice, for my data. I thought that, perhaps, this was leading to vanishing gradients, and then `nan` values propagating through the whole network. 

Since working on this, I've come to understand that the embedding layer sits outside the LSTM and so is 'exposed' to the vanishing/exploding gradients issue in a way that the gradients within an LSTM cell are not. This realization made my hunch feel more solid. I guess when one is more proficient at the underlying concepts (and python programming) the embedding layer sitting outside of the LSTM cell is obvious, but it wasn't to me at the time.

So what I did was adjust two defaults when insantiating my estimator: 
(1) I reduced the maximum embedding dimension that any of the static categorical variables could take
(2) I reduced the default value for gradient clipping

I'm not sure (2) is the best thing, since this value affects all the gradients, and not just those related to the embedding layers. Regardless, this seems to have worked in my case. Of course that doesn't mean my conceptual understanding actually explains what was going on. I could still be misunderstanding important issues. But somehow I seem to have gotten to a workable end point. 

It would be helpful if the code performed a check for 'nan' values at the point that `self.embedder(feat_static_cat)` within `prepare_rnn_input()`. For less experienced people like me that would have made it a lot easier to at least locate the origin of the issue. I got very lost trying to work backwards from an error message about degrees of freedom.

**Comment by satyrmipt** [comment]
Running into same issue with iTransformer model. The problem happens when using GPU and vanishes when only CPU is available. BTW the same code does not work on TPU too but with another error.

**Comment by satyrmipt** [comment]
@Serendipity31, @santoshkumarradha, do you managed to solve the issue? Vanishing\exploding gradients can hardly be the reason because toy dataset with constant values for every item_id work well. Different behavior on CPU, GPU and TPU also looks strange for gradients problem.
